### PR TITLE
Add ability create custom manifest operation IDs

### PR DESCRIPTION
**Pull request by jerelmiller**
The `generate-persisted-query-manifest` CLI now has the ability to generate custom manifest operation IDs. When using a config file format that supports functions (i.e. `js` and `ts` extensions), a `createCustomId` function can be added to customize how the ID should be generated for a given query.

Here is an example that shows how to create a base64 encoded ID of the query when its operation name is `TestOperation`:

```ts
// persisted-query-manifest.config.ts
import { PersistedQueryManifestConfig } from '@apollo/generate-persisted-query-manifest';
import { Buffer } from 'node:buffer'

const config: PersistedQueryManifestConfig = {
  createOperationId(query, { operationName, createDefaultId }) {
    switch (operationName) {
      case 'TestOperation':
        return Buffer.from(query).toString('base64');
      default:
        return createDefaultId();
    }
  }
}

export default config;
```

The `id` returned from `createOperationId` will be validated for uniqueness before the manifest file is written.
Screenshot 2023-06-14 at 1 16 10 AM

---

The `createOperationId` function signature is as-follows:

```ts
createOperationId(query: string, options: CreateOperationIdOptions) => string

type CreateOperationIdOptions {
  operationName: string;
  type: OperationType;
  createDefaultId: () => string;
}
```

### CODE REVIEW

File: packages/generate-persisted-query-manifest/src/index.ts [hunk 1]
```diff
@@ -101,6 +117,19 @@ const ERROR_MESSAGES = {
...
+      id,
+    )}) for operation named "${COLORS.name(
+      operationName,
+    )}" that has already been used for operation named "${COLORS.name(
+      definedOperationName,
+    )}".`;
+  },
 };
 
 function addError(source: DocumentSource, message: string) {
```
Comment (phryneas) [inline packages/generate-persisted-query-manifest/src/index.ts:135]:
```suggestion
function addError(
  source: Pick<DocumentSource, "file"> & Partial<DocumentSource>,
  message: string,
) {
```

File: packages/generate-persisted-query-manifest/src/index.ts [hunk 2]
```diff
@@ -133,15 +162,36 @@ function getDocumentSources(filepath: string): DocumentSource[] {
...
-  const sources = [...new Set(filepaths)].flatMap(getDocumentSources);
+  const {
+    documents = defaults.documents,
+    documentIgnorePatterns = defaults.documentIgnorePatterns,
+    createOperationId = defaults.createOperationId,
+  } = config;
+
+  const configFile = configFilePath
+    ? vfile({ path: relative(process.cwd(), configFilePath) })
+    : null;
```
Comment (phryneas) [inline packages/generate-persisted-query-manifest/src/index.ts:192]:
```suggestion
  const configFile = vfile({
    path: configFilePath
      ? relative(process.cwd(), configFilePath)
      : "general error",
  });
```

File: packages/generate-persisted-query-manifest/src/index.ts [hunk 3]
```diff
@@ -208,17 +254,39 @@ export async function generatePersistedQueryManifest(
...
+      // guaranteed to create unique IDs.
+      if (configFile && manifestOperationIds.has(id)) {
+        const message = configFile.message(
+          ERROR_MESSAGES.uniqueOperationId(
+            id,
+            name,
+            manifestOperationIds.get(id)!,
+          ),
+        );
+        message.fatal = true;
```
Comment (phryneas) [inline packages/generate-persisted-query-manifest/src/index.ts:283]:
```suggestion
      if (manifestOperationIds.has(id)) {
        addError(
          { file: configFile },
          ERROR_MESSAGES.uniqueOperationId(
            id,
            name,
            manifestOperationIds.get(id)!,
          ),
        );
```

File: packages/generate-persisted-query-manifest/src/index.ts [hunk 4]
```diff
@@ -229,6 +297,10 @@ export async function generatePersistedQueryManifest(
     }
   }
 
+  if (configFile) {
+    maybeReportErrorsAndExit(configFile);
+  }
```
Comment (phryneas) [inline packages/generate-persisted-query-manifest/src/index.ts:302]:
```suggestion
  maybeReportErrorsAndExit(configFile);
```

File: packages/generate-persisted-query-manifest/src/index.ts (hunk 1 above)
Comment (phryneas) [reply packages/generate-persisted-query-manifest/src/index.ts:135]: This way we can use `addError` further down without having to essentially duplicate it's logic.

**Review by phryneas — COMMENTED** [review]
I'd suggest to drop the "has a config file" specific error tracking.

Instead:
* always have a `configFile` vfile (with a real or fake path)
* checking for collisions doesn't hurt, even if they shouldn't be able to occur
* just always see if there are reportable errors

=> This allows us to report other errors in the future without bigger code changes.

**Review by phryneas — APPROVED** [review]
I don't have a local dataset to test this, and this has no tests (might be a good idea to add some, in general?), so all I could do is read.

From that, the code seems good - if you apply my suggested changes or not: approved.

**Comment by jerelmiller** [comment]
> I don't have a local dataset to test this

I've been testing on the [spotify demo](https://github.com/apollographql/spotify-showcase) for this. But yes, I want to add some tests soon (unfortunately there weren't any yet in this package when I started working on it).

File: packages/generate-persisted-query-manifest/src/index.ts (hunk 4 above)
Comment (jerelmiller) [reply packages/generate-persisted-query-manifest/src/index.ts]: Good call! Consolidated in [8f2d5dd](https://github.com/apollographql/apollo-utils/pull/309/commits/8f2d5dd5cb474dd07d6abb182d98a730489998b1)

File: packages/generate-persisted-query-manifest/src/index.ts (hunk 1 above)
Comment (jerelmiller) [reply packages/generate-persisted-query-manifest/src/index.ts]: I like this. Updated in [f5638a1](https://github.com/apollographql/apollo-utils/pull/309/commits/f5638a1beec1ef04530ea6ac4b4cd0d083eed17e)

File: packages/generate-persisted-query-manifest/src/index.ts (hunk 3 above)
Comment (jerelmiller) [reply packages/generate-persisted-query-manifest/src/index.ts]: Updated in [f5638a1](https://github.com/apollographql/apollo-utils/pull/309/commits/f5638a1beec1ef04530ea6ac4b4cd0d083eed17e)

**Pull request merged**